### PR TITLE
Moving data provider structs into their own crates

### LINKED ISSUE #415: Reconsider location for data struct definitions
With #405, it will no longer be necessary to keep all of the data struct definitions in a common crate.  Previously, it was required that all data providers have access to the struct definitions when deserializing data.  However, with this no longer the case, we should reconsider this design choice.

### PR DESCRIPTION
**Pull request by sffc**
Fixes #415

I put the structs into a `pub mod provider`, which is a convention we can adopt across all components.

Note that there was already a file named "provider.rs" in datetime; I renamed that file to "provider_internal.rs" before moving the structs file into "provider.rs".  Please keep this in mind when reviewing!

I made the `serde` feature optional on the struct, which means we can use the struct without serde (e.g. Static DataProvider).

### CODE REVIEW

**Review by zbraniecki — DISMISSED** [review]
lgtm!

I'd only suggest moving `provider.rs` and `provider_internal.rs` to `provider/structs.rs` and `provider/mod.rs`.

**Comment by sffc** [comment]
> lgtm!
> 
> I'd only suggest moving `provider.rs` and `provider_internal.rs` to `provider/structs.rs` and `provider/mod.rs`.

Reasons why I'd rather not use that particular naming scheme:

1. I'd like it if, by convention, data provider structs were directly inside the `provider` module of all components
2. The stuff inside internal_provider.rs was and still is crate-private
3. This file may be largely re-written as part of my work on #257 and #355..

I would be okay with some other name than `provider_internal`, but I do feel strongly that `provider` should not be redefined.

**Comment by zbraniecki** [comment]
> > lgtm!
> > I'd only suggest moving `provider.rs` and `provider_internal.rs` to `provider/structs.rs` and `provider/mod.rs`.
> 
> Reasons why I'd rather not use that particular naming scheme:
> 
>     1. I'd like it if, by convention, data provider structs were directly inside the `provider` module of all components

That seems like a very subjective preference, boiling down to `use crate::provider::structs::X; vs `use crate::provider::X`.
But sure, you can always have in `provider/mod.rs` `pub use structs::*;` and you have your goal met.

>     2. The stuff inside internal_provider.rs was and still is crate-private

That's accomplished by the model I proposed.
 
> I would be okay with some other name than `internal_provider`, but I do feel strongly that `provider` should not be redefined.

The file structure I proposed can provide what you're looking for. I'm very fond of clustering analogous mental spaces in folders.
I find it vastly easier to reason about the code when there's `src/provider` dir with modules inside, than `foo_provider.rs`, `bar_provider.rs`, `provider.rs` and `baz_provider.rs` in a `src` directory mixed with 5-10 other files.

How you expose those is flexible, but I also prefer `use crate::provider::foo;` and `use crate::provider::bar` over `use crate::internal_provider;` and `use crate::provider;`.

**Comment by sffc** [comment]
Are you okay with provider/mod.rs and provider/helpers.rs?

**Comment by sffc** [comment]
The merge directly from master is confusing, so I'm squashing and rebasing instead.  ^ that was the squash.  Rebase coming up next.

**Comment by sffc** [comment]
This is done except for the features test breakage.

The problem is that `icu_locale_canonicalizer` has an optional `serde` dependency, but that dependency requires that the `icu_locid/serde` feature is also enabled.  As far as I can tell, it's not possible to make optional dependencies enable features of other dependencies.

With a normal feature (not an optional dependency), you can do:

```
[features]
foo = ["bar/baz"]
```

@Manishearth Suggestions?

**Comment by sffc** [comment]
I'm leaning toward adding a new feature that can pull in `serde` and whatever else is needed, and using this same feature name across all the ICU4X components.

What should we call the new feature?

- `serialize`
- `deserialize`
- `serialize_deserialize`
- `icu4x_serde`
- `provider_serde`
- `enable_serde`
- other?

**Comment by zbraniecki** [comment]
+1 for "serialize"

**Comment by Manishearth** [comment]
serialize seems fine to me. Not happy about having to do this, but all we can do is wait for namespaced features I guess

**Comment by sffc** [comment]
I went with `provider_serde` because I realized that we may want the data provider serialization to have a separate toggle switch from serialization in other parts of the library (such as LanguageIdentifier).

Looks like the tests are all passing, including the feature permutation test!  This is ready for re-review, and I'm eager to check it in.  I may wait to hit the merge button until tomorrow's meeting.

**Review by echeran — APPROVED** [review]
Looks good for `provider_ppucd` and `uniset` components, but of course not much needed coincidentally.

File: components/provider/src/hello_world.rs
Comment (zbraniecki) [inline components/provider/src/hello_world.rs:82]: Suggestion:
```rust
let data = [
    ("bn", "ওহে বিশ্ব"),
    ("cs", "Ahoj světe"),
].iter().map(|(loc, value)| {
    (loc.parse().unwrap(), Cow::Borrowed(value))
}).collect();
```

**Review by zbraniecki — DISMISSED** [review]
lgtm! Thank you Shane!

File: components/provider/src/hello_world.rs
Comment (sffc) [reply components/provider/src/hello_world.rs]: Done!

**Review by zbraniecki — DISMISSED** [review]

**Comment by sffc** [comment]
Clippy seems to be failing due to the known clippy bug:

```
error: field assignment outside of initializer for an instance created with Default::default()
   --> components/provider_fs/src/bin/icu4x-cldr-export.rs:273:5
    |
273 |     options.root = output_path;
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: `-D clippy::field-reassign-with-default` implied by `-D warnings`
note: consider initializing the variable with `icu_provider_fs::export::fs_exporter::ExporterOptions { root: output_path, ..Default::default() }` and removing relevant reassignments
   --> components/provider_fs/src/bin/icu4x-cldr-export.rs:272:5
    |
272 |     let mut options = fs_exporter::ExporterOptions::default();
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#field_reassign_with_default

```

So I believe this PR is ready to land as soon as I get the final approvals.

**Review by nciric — APPROVED** [review]

**Comment by sffc** [comment]
@filmil told me offline that he's unable to review this PR, so I will proceed to merge with the approvals from @zbraniecki, @nciric, and @echeran.

**Pull request merged**
